Add a skill for benchmarking and switch to GHCP SDK instead of Inspect AI (#19)
microsoft/a11y-llm-eval: config/instructions/accessible-basic.md

---
description: "Accessibility coding rules: WCAG 2.2 AA conformance, semantic structure, keyboard support, focus management. Apply when writing or modifying UI code."
applyTo: "**"
---

# Accessibility instructions (standard)

Conform to [WCAG 2.2 Level AA](https://www.w3.org/TR/WCAG22/). Do not claim output is "fully accessible" — state what was addressed and any known limitations.

## Implementation priority

Use the first option that fits:

1. Existing accessible component in the project / design system.
2. A component library already in use.
3. Native platform semantics (`<button>`, `<a href>`, `<input>`, `<label>`, `<fieldset>`/`<legend>`, `<dialog>`, `<nav>`, `<main>`, `<h1>`–`<h6>`).
4. Native element + minimum necessary ARIA.
5. Fully custom ARIA widget — only when nothing above fits, and only with complete APG keyboard, focus, and state behavior.

> No ARIA is better than bad ARIA. Don't duplicate native semantics (no `role="button"` on `<button>`). Don't use `role="menu"` for site navigation.

Reviewer: ```suggestion
No ARIA is better than bad ARIA. Don't duplicate native semantics (no `role="button"` on `<button>`). Don't use `role="menu"` for site navigation. Don't use `aria-checked` on native checkboxes. etc.
```

Author: I left this out on purpose - trying to give hints/examples but not be comprehensive.

## Structure

- Use landmark elements (`<header>`, `<nav>`, `<main>`, `<footer>`). Exactly one `<main>`. Give duplicated landmarks unique accessible names.
- One `<h1>` per view, typically first heading in `<main>`. Don't skip heading levels.
- Set a descriptive `<title>`.

> - **Web pages only:** Provide a "Skip to main content" link as the first focusable element (not required for SPAs or Electron).

Reviewer: i think there's an argument to be made that skip to main content links still have usefulness for SPAs / Electron.

Author: I had removed SPAs. Not sure how it got back in there. I'm less sure about electron, but I'll remove that reference for now.

## Name, role, value

- Every interactive element exposes an accurate accessible name. Role matches purpose. Dynamic states (pressed, expanded, selected, checked, disabled, invalid) stay in sync with visuals.

> - The accessible name MUST contain the visible label text. When multiple controls share a label (e.g. many "Remove" buttons), add context: `aria-label="Remove item: Socks"`.

Reviewer: ```suggestion
- The accessible name MUST contain the visible label text. When multiple controls share a label (e.g., many "Remove" buttons), add context: `aria-label="Remove item: Socks"`.
```

## Keyboard and focus

> - All interactive elements are keyboard operable; tab order matches reading/visual order.

Reviewer: not every interactive element needs to be keyboard accessible.  but all interactivity needs to be achievable by use of keyboard :)

```suggestion
- All interactive functionality can be achieved by both mouse an keyboard; tab order matches reading/visual order.
```

> - Focus is always visible — do not remove focus outlines without an equal-or-better replacement.

Reviewer: ```suggestion
- Focus is always visible — do not remove focus outlines without an equal-or-better replacement. Box-shadow alone is not sufficient since it is not a respected style in high contrast themes.
```

Author: I don't think this is necessary for the basic instructions.

> - No keyboard traps outside modals. Escape closes overlays.

Reviewer: we should talk about this one

Author: I made an update to this to make it more accurate.

- Static content MUST NOT be sequentially focusable. Use `tabindex="-1"` only for programmatic focus targets.

> - Content hidden from AT (`hidden`, `display:none`, `aria-hidden="true"`) MUST NOT be focusable.

Reviewer: display none / hidden content can't be focusable.  so that part of this statement doesn't make sense. how about just:
```suggestion
- Content hidden from AT with `aria-hidden="true"` MUST NOT be focusable.
```

Author: Fair enough

> - Overlays/dialogs/disclosures move focus in and restore it on close.

Reviewer: mmm..... not always?  especially not sure why disclosures are here.. unless this means like popup/popover disclosures?  but even then...

- Composite widgets (tabs, listbox, menu, grid): one tab stop total; arrow keys move focus internally via roving `tabindex` or `aria-activedescendant`.

## Forms

> - Every control has a visual and programmatic label (`<label for>` or wrapping `<label>`). Never rely on placeholder alone.

Reviewer: every "form field"?  "control" could be buttons as well, and buttons can't have placeholders and commonly don't need labels.

- Associate help/error text via `aria-describedby`.
- Group related options (checkboxes, radios) with `<fieldset>` + `<legend>`.

> - Required fields: visible indicator (e.g. `*`) AND `required` / `aria-required="true"`. Never color alone.

Reviewer: i think this might need to have the nuance for forms with required fields AND optional fields, provide a visual indicator.....

since login form of user name and password should not need * to indicate they are required

- Invalid fields: `aria-invalid="true"`; remove when corrected. Error messages explain how to fix.
- On submit with invalid input, focus the first invalid control. Don't disable submit solely to prevent submission.

## Contrast and color

- Text contrast ≥ 4.5:1 (≥ 3:1 for large text: 24px regular or 18.66px bold).
- Focus indicators and key control boundaries ≥ 3:1 vs adjacent colors.
- Never use color as the only cue for meaning (error, success, required, selected).
- Use design tokens / CSS custom properties. Avoid `opacity`, `rgba`, `hsla` on text and essential affordances — contrast becomes background-dependent.
- Ensure contrast in all states: default, hover, active, focus, visited, disabled.

## Forced colors / OS settings

> - Never override OS high-contrast, reduced-motion, or color-scheme preferences. Do not use `forced-color-adjust: none`.

Reviewer: >Do not use `forced-color-adjust: none`

there are times when you do want to use this.  e.g., data viz where we _want_ the colors to remain unchanged

- In `@media (forced-colors: active)`, use system color keywords (`ButtonText`, `ButtonBorder`, `CanvasText`, `Canvas`) — never fixed hex/RGB.
- Use `currentColor` for SVG `fill`/`stroke` so icons inherit the foreground.
- If relying on `box-shadow` for focus, add a transparent `outline` so focus renders in forced colors.

## Reflow (SC 1.4.10)

> - Content MUST adapt to 320 CSS pixels wide without two-dimensional scrolling for multi-line text.

Reviewer: ```suggestion
- Content MUST be able to adapt to 320 CSS pixels wide without two-dimensional scrolling for multi-line text.
```

> - Multi-column layouts stack; text wraps; controls remain operable.

Reviewer: ```suggestion
- For multi-column layouts that are not necessary to convey meaning or important to the UX of the interface, content stacks; text wraps; controls remain operable.
```

- Use fluid `flex`/`grid`. Set `max-width: 100%` on media, `min-width: 0` on flex/grid children, `overflow-wrap: anywhere` for long strings.

> - Exception: inherently 2D components (large tables, maps, charts) may scroll horizontally at component level; the surrounding view still reflows.

Reviewer: ```suggestion
- Exception: inherently 2D components (large tables, maps, charts, media, interfaces with toolbars or interfaces that require 2D layout) may scroll horizontally at component level; the surrounding view still reflows.
```

## Graphics

- Informative `<img>` → meaningful `alt`. Decorative `<img>` → `alt=""`.
- Informative `<svg>` → `role="img"` with `aria-label` / `aria-labelledby`. Decorative SVG/graphics → `aria-hidden="true"`.

## Navigation

- Use `<nav>` with lists and links — not `role="menu"` / `role="menubar"`.

> - Expandable menus: toggle uses `aria-expanded`. Escape MAY close sub-navigations.

Reviewer: probably shouldn't call them expandable menus when having just stated to not use menus :D

also, probably need to clarify what element the aria-expanded goes on.  e.g., a button element.

## Tables and grids

- Static tabular data: `<table>` with `<th>` for column/row headers.
- Use `role="grid"` only for genuinely interactive tabular UIs, with proper row/cell nesting and arrow-key navigation.

## Status messages

> - Announce dynamic updates (loading, submission results, validation summaries) via `aria-live="polite"` or `aria-live="assertive"`.

Reviewer: i'm worried "submission results" would make one think that all results from a search submission need to be announced.

and i say this because i've seen LLMs output entire pages in a live region because CONTENT NEED TO BE ANNOUNCED LOLOLOL
